api/_db.data-utils.js

```diff
@@ -0,0 +1,127 @@
+	assert(parsed instanceof Date && !isNaN(parsed.valueOf()), `Invalid Date: "${inspect(parsed)}"`);
+};
+
+const formatDate = (parsed) => {
+	if (!parsed) {
+		return null;
+	}
+	assertValidDate(parsed);
+	return parsed.toISOString().substr(0, 10);
+};
+
+const parseDate = (rawValue) => {
+	if (!rawValue) {
+		return null;
+	}
+	const parsed = new Date(rawValue);
+	assert(!isNaN(parsed), `Invalid date: "${inspect(rawValue)}"`);
+	return parsed;
+};
+
+const S = 1000;
+const MIN = 60 * S;
+const HR = 60 * MIN;
+const D = 24 * HR;
+const getExpiryTimeFromFormFillTime = (normDate) => {
+	if (!normDate) {
+		return null;
+	}
+	assertValidDate(normDate);
+	// Add 6 weeks
+	const exp = normDate.valueOf() + 6 * 7 * D;
+	return new Date(exp);
+};
+
+// have to either map by the position or header name, doing the latter
+const filterColumnHeader = 'id';
+const certificateHeader = 'certificate';
+const nameHeader = 'name';
+const examinerHeader = 'examiner';
+const examDateHeader = 'examDate';
+const expiryDateHeader = 'expiryDate';
+const formFillTimeHeader = 'formFillTime';
+
+export const findById = (data, id) => {
+	// Drop human-readable headers
+	data.shift();
+	const headers = data.shift();
+
+	const filterColumnIdx = headers.indexOf(filterColumnHeader);
+	const certificateColumnIdx = headers.indexOf(certificateHeader);
+	const nameColumnIdx = headers.indexOf(nameHeader);
+	const examinerColumnIdx = headers.indexOf(examinerHeader);
+	const examDateColumnIdx = headers.indexOf(examDateHeader);
+	const expiryDateColumnIdx = headers.indexOf(expiryDateHeader);
+	const formFillTimeColumnIdx = headers.indexOf(formFillTimeHeader);
+
+	assert(~filterColumnIdx, `Filter column not found. Looked for "${filterColumnHeader}"`);
+	assert(~certificateColumnIdx, `Certificate column not found. Looked for "${certificateHeader}"`);
+	assert(~nameColumnIdx, `Certificate column not found. Looked for "${nameHeader}"`);
+	assert(~examinerColumnIdx, `Examiner column not found. Looked for "${examinerHeader}"`);
+	assert(~examDateColumnIdx, `Exam time column not found. Looked for "${examDateHeader}"`);
+	assert(~expiryDateColumnIdx, `Expiry time column not found. Looked for "${expiryDateHeader}"`);
+	if (formFillTimeColumnIdx === -1) {
+		console.warn(`Form fill time column not found. Looked for "${formFillTimeHeader}"`);
+	}
+
+	const filteredRows = data.filter((row) => {
+		return row[filterColumnIdx] === id;
+	});
+
+	// no result
+	assert(filteredRows.length, 'Not found');
+
+	// more than one result, assume an filtering error
+	if (filteredRows.length > 1) {
+		console.log('Filtered more than one row', filteredRows);
+		throw new Error('More than one row');
+	}
+
+	const row = filteredRows[0];
+	const certificate = normalizeCertificate(row[certificateColumnIdx] ?? '');
+
+	const formFillDate = parseDate(row[formFillTimeColumnIdx]);
+	const expiryDate = parseDate(row[expiryDateColumnIdx]) || getExpiryTimeFromFormFillTime(formFillDate) || null;
```

Comment:
What happens when neigther expiryDate or expiryTimeFromFormFillTime are set?

Reply:
Then `null` is returned here and the UI shows it as expired.

What should happen?

Reply:
The current logic is ok.